src/mercury_engine_data_structures/formats/bmsas.py

```diff
@@ -16,91 +16,123 @@
     Switch,
 )
 
-from mercury_engine_data_structures.common_types import DictAdapter, Float, make_vector
+from mercury_engine_data_structures.common_types import Char, DictAdapter, Float, make_vector
+from mercury_engine_data_structures.common_types import StrId as StrIdSR
 from mercury_engine_data_structures.construct_extensions.strings import PascalStringRobust
 from mercury_engine_data_structures.formats import BaseResource
-from mercury_engine_data_structures.formats.property_enum import PropertyEnum
+from mercury_engine_data_structures.formats.property_enum import PropertyEnum, PropertyEnumSR
 from mercury_engine_data_structures.game_check import Game
 
 StrId = PascalStringRobust(Int16ul, "utf-8")
 
 
 class StrIdOrInt(Select):
-    def __init__(self):
-        super().__init__(StrId, Hex(Int64ul))
+    def __init__(self, str_subcon, int_subcon):
+        self.str_subcon = str_subcon
+        self.int_subcon = int_subcon
+        super().__init__(self.str_subcon, Hex(self.int_subcon))
 
     def _emitparse(self, code: construct.CodeGen):
         code.append(f"""
             def parse_str_or_int(io):
                 fallback = io.tell()
                 try:
-                    return {StrId._compileparse(code)}
+                    return {self.str_subcon._compileparse(code)}
                 except UnicodeDecodeError:
                     io.seek(fallback)
-                    return {Int64ul._compileparse(code)}
+                    return {self.int_subcon._compileparse(code)}
         """)
         return "parse_str_or_int(io)"
 
     def _emitbuild(self, code: construct.CodeGen):
-        return f"({Int64ul._compilebuild(code)}) if isinstance(obj, int) else ({StrId._compilebuild(code)})"
+        return (
+            f"({self.int_subcon._compilebuild(code)})"
+            "if isinstance(obj, int) else "
+            f"({self.str_subcon._compilebuild(code)})"
+        )
+
+ArgTypes = {
+    'f': Float,
+    'b': Flag,
+    'u': Int32ul,
+    'i': Int32sl,
+    'e': Int32ul,
+    'o': Int32ul,
+    't': Int32ul,
+}
+
+ArgTypesDread = {
+    **ArgTypes,
+    's': StrIdOrInt(StrId, Int64ul),
+    'v': Switch(
+        construct.this.key,
+        {
+            'vColorStart': Int32ul,
+            'vColorEnd': Int32ul,
+        },
+        default=Array(3, Float)
+    ),
+}
+
+ArgTypesSR = {
+    **ArgTypes,
+    's': StrIdOrInt(StrIdSR, Int32ul),
+}
 
 
-Argument = Struct(
+ArgListDread = DictAdapter(make_vector(Struct(
```

Comment:
Make a `_arg_list` function that creates the `ArgListXX`, since it only differs by the `ArgTypes`

Reply:
no, the value differs as well. SR has an explicit type field which dread lacks